lib/schema/union.js

```diff
@@ -67,16 +115,44 @@ class Union extends SchemaType {
           castedVal = this.schemaTypes[i].cast(castedVal, doc, init, prev, options);
         }
         if (castedVal === val) {
+          if (castedVal != null && typeof castedVal === 'object' && castedVal.$__ != null) {
+            castedVal.$__[castSchemaTypeSymbol] = this.schemaTypes[i];
+          }
           return castedVal;
         }
+
+        if (isObject && castedVal != null && typeof castedVal === 'object') {
+          const inputKeys = Object.keys(val);
+          const preservedFields = inputKeys.filter(key => key in castedVal && key !== '_id').length;
+          const score = preservedFields;
+
+          if (score > bestMatchScore) {
```

Comment:
Going for a best match scoring is a bit overkill IMO, this sort of pattern will be difficult to explain and document. I would prefer to keep more of a "first entry in the union that matches" approach.